File: app/src/main/java/com/github/goldy1992/mp3player/dagger/components/MediaActivityCompatComponent.java
package com.github.goldy1992.mp3player.dagger.components;

import android.content.Context;

import com.github.goldy1992.mp3player.client.MediaBrowserConnectorCallback;
import com.github.goldy1992.mp3player.client.activities.FolderActivity;
import com.github.goldy1992.mp3player.client.activities.MainActivity;
import com.github.goldy1992.mp3player.client.activities.MediaPlayerActivity;
import com.github.goldy1992.mp3player.client.views.fragments.PlaybackSpeedControlsFragment;
import com.github.goldy1992.mp3player.dagger.components.fragments.ChildViewPagerFragmentSubcomponent;
import com.github.goldy1992.mp3player.dagger.components.fragments.PlaybackButtonsSubComponent;
import com.github.goldy1992.mp3player.dagger.components.fragments.PlaybackTrackerFragmentSubcomponent;
import com.github.goldy1992.mp3player.dagger.modules.ComponentNameModule;
import com.github.goldy1992.mp3player.dagger.modules.GlideModule;
import com.github.goldy1992.mp3player.dagger.modules.MainHandlerModule;
import com.github.goldy1992.mp3player.dagger.modules.MediaBrowserCompatModule;
import com.github.goldy1992.mp3player.dagger.modules.MyDrawerListenerModule;
import com.github.goldy1992.mp3player.dagger.modules.service.HandlerThreadModule;
import com.github.goldy1992.mp3player.dagger.scopes.ComponentScope;

import dagger.BindsInstance;
import dagger.Component;

@ComponentScope
@Component(modules = {
        GlideModule.class,
        ComponentNameModule.class,
        HandlerThreadModule.class,
        MainHandlerModule.class,
        MediaBrowserCompatModule.class,
        MyDrawerListenerModule.class
        })
public interface MediaActivityCompatComponent {

    // activities
    void inject(MainActivity mainActivity);
    void inject(MediaPlayerActivity mediaPlayerActivity);
    void inject(FolderActivity folderActivity);

    // fragments
    void inject(PlaybackSpeedControlsFragment playbackSpeedControlsFragment);

    // sub components
    ChildViewPagerFragmentSubcomponent.Factory childViewPagerFragmentSubcomponentFactory();
    PlaybackTrackerFragmentSubcomponent playbackTrackerSubcomponent();
    PlaybackButtonsSubComponent playbackButtonsSubcomponent();
    SearchResultActivitySubComponent searchResultActivitySubComponent();
    SplashScreenEntryActivityComponent.Factory splashScreenEntryActivity();


    @Component.Factory
    interface Factory {
        MediaActivityCompatComponent create(@BindsInstance Context context,
                                            @BindsInstance String workerId,
                                            @BindsInstance MediaBrowserConnectorCallback callback);
    }
}

File: client/src/main/java/com/github/goldy1992/mp3player/client/dagger/subcomponents/MediaActivityCompatComponent.kt
package com.github.goldy1992.mp3player.client.dagger.subcomponents

import android.content.Context
import com.github.goldy1992.mp3player.client.activities.*
import com.github.goldy1992.mp3player.client.dagger.modules.*
import com.github.goldy1992.mp3player.commons.ComponentClassMapper
import com.github.goldy1992.mp3player.commons.dagger.scopes.ComponentScope
import dagger.BindsInstance
import dagger.Component

@ComponentScope
@Component(modules = [
    GlideModule::class,
    MediaBrowserAdapterModule::class,
    MediaControllerAdapterModule::class,
    MediaBrowserCompatModule::class,

    // supcomponents module
    MediaActivitySubcomponents::class
])
interface MediaActivityCompatComponent {

    /* Activities */
    fun inject(folderActivity: FolderActivity)
    fun inject(mainActivity: MainActivity)
    fun inject(mediaPlayerActivity: MediaPlayerActivity)
    fun inject(searchResultActivity: SearchResultActivity)
    fun inject(mediaActivityCompat: MediaActivityCompat)

    /* Subcomponents */
    fun mediaFragmentSubcomponent() : MediaFragmentSubcomponent.Factory
    fun mediaItemListFragmentSubcomponent() : MediaItemListFragmentSubcomponent.Factory

    @Component.Factory
    interface Factory {
        fun create(@BindsInstance context: Context,
                   @BindsInstance componentClassMapper: ComponentClassMapper)
                : MediaActivityCompatComponent
    }
}